Add `CanvasItem.draw_nine_patch()` method

dalexeev opened the pull request:
Closes godotengine/godot-proposals#6746.

dalexeev committed:
Add `CanvasItem.draw_nine_patch()` method

YuriSizov commented:
9 arguments feels a bit too much for a high-level user facing method. What's the most basic use case for calling this? I would design the `CanvasItem` API around that, rather than making a thin wrapper over the server method.

KoBeWi commented:
`topleft` and `bottomright` can be turned into a rect.
Also `texture` should be the first argument, to be consistent with other drawing methods.
`source` can be optional argument. When you draw ninepatch you usually use the whole texture.

dalexeev commented:
I agree with you, but we already have this:

* `draw_multiline_string_outline` - 13 parameters
* `draw_multiline_string` - 12 parameters
* `draw_string_outline` - 11 parameters
* `draw_string` - 10 parameters

Even the thin wrapper makes sense, as users are more familiar with `draw_*` methods than `RenderingServer`, see the linked discussion.

Also, we could use `NINE_PATCH_TILE` instead of `NINE_PATCH_STRETCH` for `x_axis_mode` and `y_axis_mode`. Perhaps we shouldn't remove anything, just change the order and add defaults.

> `topleft` and `bottomright` can be turned into a rect.

`Vector2(10, 10), Vector2(20, 20)` makes more sense than `Rect2(10, 10, texture.get_width() - 10 - 20, texture.get_height() - 10 - 20)`.

YuriSizov commented:
> I agree with you, but we already have this:
> 
> * `draw_multiline_string_outline` - 13 parameters
> * `draw_multiline_string` - 12 parameters
> * `draw_string_outline` - 11 parameters
> * `draw_string` - 10 parameters

Well, IMO these are also problematic :) I think a thin wrapper doesn't improve usability, it just makes the method slightly more discoverable, but still hard to use and read and parse mentally. Even reading the docs can be daunting for those. We should've simplified the high-level methods to the most basic case and linked to the rendering server for more advanced cases, but alas the time has passed for that.

Either way, yes, we should at least order the arguments to allow the simplest calls to be nice and short.

KoBeWi commented:
> Vector2(10, 10), Vector2(20, 20) makes more sense than Rect2(10, 10, texture.get_width() - 10 - 20, texture.get_height() - 10 - 20).

Well the idea is that with rect you specify the center of the texture that will fill the ninepatch. But I guess it makes setting precise margins more difficult 🤔
Another thing to consider could be making `bottomright` optional and equal to `topleft` by default.

In any case, these 2 arguments should be renamed to have underscore in name (`top_left`, `bottom_right`).

dalexeev commented:
> `source` can be optional argument. When you draw ninepatch you usually use the whole texture.

I split the method into `draw_nine_patch()` and `draw_nine_patch_region()` since there is already `draw_texture_rect()` and `draw_texture_rect_region()`.

> Also `texture` should be the first argument, to be consistent with other drawing methods.

I'm not sure about this since `rect` is a standalone parameter and everything else (including the texture) is about the ninepatch setup.

> Also, we could use `NINE_PATCH_TILE` instead of `NINE_PATCH_STRETCH` for `x_axis_mode` and `y_axis_mode`.

Now I think this is a bad idea because both the `RenderingServer` method and the `NinePatchRect` control use `STRETCH` by default.

> In any case, these 2 arguments should be renamed to have underscore in name (`top_left`, `bottom_right`).

Done.

> Another thing to consider could be making `bottomright` optional and equal to `topleft` by default.

I think this is too "magical". Perhaps we should make `Vector2(0, 0)` the default value for both arguments since `NinePatchRect` also has `patch_margin_* == 0` by default?


I also moved `draw_center` and modulate before `x_axis_mode` and `y_axis_mode`.

dalexeev commented:
This PR is outdated and I no longer have the time or desire to continue working on it. You can use this as a basis and credit me as a co-author if you want.
